table/block_based/block_based_table_builder.cc

```diff
@@ -767,34 +1033,159 @@ void BlockBasedTableBuilder::WriteRawBlock(const Slice& block_contents,
       }
     }
 
-    assert(r->io_status.ok());
+    assert(io_s.ok());
     TEST_SYNC_POINT_CALLBACK(
         "BlockBasedTableBuilder::WriteRawBlock:TamperWithChecksum",
         static_cast<char*>(trailer));
-    r->io_status = r->file->Append(Slice(trailer, kBlockTrailerSize));
-    if (r->io_status.ok()) {
-      r->status = InsertBlockInCache(block_contents, type, handle);
+    io_s = r->file->Append(Slice(trailer, kBlockTrailerSize));
+    if (io_s.ok()) {
+      s = InsertBlockInCache(block_contents, type, handle);
+      if (!s.ok()) {
+        SetStatusAtom(s);
+      }
+    } else {
+      SetIOStatusAtom(io_s);
     }
-    if (r->status.ok() && r->io_status.ok()) {
+    if (s.ok() && io_s.ok()) {
       r->offset += block_contents.size() + kBlockTrailerSize;
       if (r->table_options.block_align && is_data_block) {
         size_t pad_bytes =
             (r->alignment - ((block_contents.size() + kBlockTrailerSize) &
                              (r->alignment - 1))) &
             (r->alignment - 1);
-        r->io_status = r->file->Pad(pad_bytes);
-        if (r->io_status.ok()) {
+        io_s = r->file->Pad(pad_bytes);
+        if (io_s.ok()) {
           r->offset += pad_bytes;
+        } else {
+          SetIOStatusAtom(io_s);
+        }
+      }
+      if (r->compression_opts.parallel_threads > 1) {
+        if (!r->pc_rep->finished) {
+          double old_compression_ratio =
+              r->pc_rep->curr_compression_ratio.load(std::memory_order_relaxed);
+          double new_compression_ratio =
+              (old_compression_ratio * r->pc_rep->raw_bytes_compressed +
+                block_contents.size()) /
+              static_cast<double>(r->pc_rep->raw_bytes_compressed +
+                                  r->pc_rep->raw_bytes_curr_block);
+          r->pc_rep->curr_compression_ratio.store(new_compression_ratio,
+                                                  std::memory_order_relaxed);
+          r->pc_rep->raw_bytes_compressed += r->pc_rep->raw_bytes_curr_block;
+          uint64_t new_raw_bytes_inflight =
+              r->pc_rep->raw_bytes_inflight.fetch_sub(
+                  r->pc_rep->raw_bytes_curr_block, std::memory_order_relaxed) -
+              r->pc_rep->raw_bytes_curr_block;
+          uint64_t new_blocks_inflight = r->pc_rep->blocks_inflight.fetch_sub(
+                                             1, std::memory_order_relaxed) -
+                                         1;
+          r->pc_rep->estimated_file_size =
+              r->offset +
+              static_cast<uint64_t>(static_cast<double>(new_raw_bytes_inflight) *
+                                    new_compression_ratio) +
+              new_blocks_inflight * kBlockTrailerSize;
+        } else {
+          r->pc_rep->estimated_file_size = r->offset;
         }
       }
     }
+  } else {
+    SetIOStatusAtom(io_s);
+  }
+  if (!io_s.ok() && s.ok()) {
```

Comment:
In origin code, `rep_->status` is overwritten by `rep_->io_status` even when the former is non-OK. I think we should expose the first failure occurred, so I changed it to only overwrite `rep_->status` with error `io_s` when the former is clean.